### PR TITLE
AV-76498 | manage GSI indexes

### PR DESCRIPTION
**Pull request by l0n3star**
## Jira

* [AV-76498](https://jira.issues.couchbase.com/browse/AV-76498)

## Description

Feature to manage primary and secondary indexes.

Design doc:
https://docs.google.com/document/d/1pTjOIA_DuvEkpj83QQC-9VRYFuOUhUcy-NIOTYh4mVM/edit?usp=sharing

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [X] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

### How was this change tested and do you have evidence? _**(REQUIRED: Select at least 1)**_

- [X] Manually tested
- [ ] Unit tested
- [ ] Acceptance tested
- [ ] Unable to test / will not test (Please provide comments in section below)

### Testing

<details open>
  <summary>Testing</summary>
  
// create new indexes

```
terraform apply
╷
│ Warning: Provider development overrides are in effect
│
│ The following provider development overrides are set in the CLI configuration:
│  - couchbasecloud/couchbase-capella in /Users/$USER/GolandProjects/terraform-provider-couchbase-capella/bin
│
│ The behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published releases.
╵

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # couchbase-capella_query_indexes.new_indexes["idx1"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field1",
        ]
      + index_name      = "idx1"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build   = false
          + num_partition = 8
          + num_replica   = 0
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx2"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field2",
        ]
      + index_name      = "idx2"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build   = false
          + num_partition = 8
          + num_replica   = 0
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx3"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field3",
        ]
      + index_name      = "idx3"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build   = false
          + num_partition = 8
          + num_replica   = 0
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx4"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field5",
        ]
      + index_name      = "idx4"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build   = false
          + num_partition = 8
          + num_replica   = 0
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx5"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field5",
        ]
      + index_name      = "idx5"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build   = false
          + num_partition = 8
          + num_replica   = 0
        }
    }

Plan: 5 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

couchbase-capella_query_indexes.new_indexes["idx2"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx5"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx3"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx1"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx4"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx3"]: Creation complete after 4s
couchbase-capella_query_indexes.new_indexes["idx2"]: Creation complete after 9s
couchbase-capella_query_indexes.new_indexes["idx5"]: Still creating... [10s elapsed]
couchbase-capella_query_indexes.new_indexes["idx1"]: Still creating... [10s elapsed]
couchbase-capella_query_indexes.new_indexes["idx4"]: Still creating... [10s elapsed]
couchbase-capella_query_indexes.new_indexes["idx1"]: Creation complete after 14s
couchbase-capella_query_indexes.new_indexes["idx4"]: Creation complete after 19s
couchbase-capella_query_indexes.new_indexes["idx5"]: Still creating... [20s elapsed]
couchbase-capella_query_indexes.new_indexes["idx5"]: Creation complete after 24s

Apply complete! Resources: 5 added, 0 changed, 0 destroyed.
```

// drop indexes

```
terraform apply
╷
│ Warning: Provider development overrides are in effect
│
│ The following provider development overrides are set in the CLI configuration:
│  - couchbasecloud/couchbase-capella in /Users/$USER/GolandProjects/terraform-provider-couchbase-capella/bin
│
│ The behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published releases.
╵
data.couchbase-capella_query_indexes.list: Reading...
couchbase-capella_query_indexes.new_indexes["idx5"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx3"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx4"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx2"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx1"]: Refreshing state...
data.couchbase-capella_query_indexes.list: Read complete after 0s

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # couchbase-capella_query_indexes.new_indexes["idx1"] will be destroyed
  # (because couchbase-capella_query_indexes.new_indexes is not in configuration)
  - resource "couchbase-capella_query_indexes" "new_indexes" {
      - bucket_name     = "test" -> null
      - cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - collection_name = "test" -> null
      - index_keys      = [
          - "field1",
        ] -> null
      - index_name      = "idx1" -> null
      - organization_id = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - project_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - scope_name      = "test" -> null
      - with            = {
          - defer_build   = false -> null
          - num_partition = 8 -> null
          - num_replica   = 0 -> null
        } -> null
    }

  # couchbase-capella_query_indexes.new_indexes["idx2"] will be destroyed
  # (because couchbase-capella_query_indexes.new_indexes is not in configuration)
  - resource "couchbase-capella_query_indexes" "new_indexes" {
      - bucket_name     = "test" -> null
      - cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - collection_name = "test" -> null
      - index_keys      = [
          - "field2",
        ] -> null
      - index_name      = "idx2" -> null
      - organization_id = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - project_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - scope_name      = "test" -> null
      - with            = {
          - defer_build   = false -> null
          - num_partition = 8 -> null
          - num_replica   = 0 -> null
        } -> null
    }

  # couchbase-capella_query_indexes.new_indexes["idx3"] will be destroyed
  # (because couchbase-capella_query_indexes.new_indexes is not in configuration)
  - resource "couchbase-capella_query_indexes" "new_indexes" {
      - bucket_name     = "test" -> null
      - cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - collection_name = "test" -> null
      - index_keys      = [
          - "field3",
        ] -> null
      - index_name      = "idx3" -> null
      - organization_id = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - project_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - scope_name      = "test" -> null
      - with            = {
          - defer_build   = false -> null
          - num_partition = 8 -> null
          - num_replica   = 0 -> null
        } -> null
    }

  # couchbase-capella_query_indexes.new_indexes["idx4"] will be destroyed
  # (because couchbase-capella_query_indexes.new_indexes is not in configuration)
  - resource "couchbase-capella_query_indexes" "new_indexes" {
      - bucket_name     = "test" -> null
      - cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - collection_name = "test" -> null
      - index_keys      = [
          - "field5",
        ] -> null
      - index_name      = "idx4" -> null
      - organization_id = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - project_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - scope_name      = "test" -> null
      - with            = {
          - defer_build   = false -> null
          - num_partition = 8 -> null
          - num_replica   = 0 -> null
        } -> null
    }

  # couchbase-capella_query_indexes.new_indexes["idx5"] will be destroyed
  # (because couchbase-capella_query_indexes.new_indexes is not in configuration)
  - resource "couchbase-capella_query_indexes" "new_indexes" {
      - bucket_name     = "test" -> null
      - cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - collection_name = "test" -> null
      - index_keys      = [
          - "field5",
        ] -> null
      - index_name      = "idx5" -> null
      - organization_id = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - project_id      = "ffffffff-aaaa-1414-eeee-000000000000" -> null
      - scope_name      = "test" -> null
      - with            = {
          - defer_build   = false -> null
          - num_partition = 8 -> null
          - num_replica   = 0 -> null
        } -> null
    }

Plan: 0 to add, 0 to change, 5 to destroy.

Changes to Outputs:
  + list_indexes = {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + data            = [
          + {
              + definition = "CREATE INDEX `idx1` ON `test`.`test`.`test`(`field1`)"
              + index_name = "idx1"
            },
          + {
              + definition = "CREATE INDEX `idx2` ON `test`.`test`.`test`(`field2`)"
              + index_name = "idx2"
            },
          + {
              + definition = "CREATE INDEX `idx3` ON `test`.`test`.`test`(`field3`)"
              + index_name = "idx3"
            },
          + {
              + definition = "CREATE INDEX `idx4` ON `test`.`test`.`test`(`field5`)"
              + index_name = "idx4"
            },
          + {
              + definition = "CREATE INDEX `idx5` ON `test`.`test`.`test`(`field5`)"
              + index_name = "idx5"
            },
        ]
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
    }

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

couchbase-capella_query_indexes.new_indexes["idx1"]: Destroying...
couchbase-capella_query_indexes.new_indexes["idx3"]: Destroying...
couchbase-capella_query_indexes.new_indexes["idx5"]: Destroying...
couchbase-capella_query_indexes.new_indexes["idx2"]: Destroying...
couchbase-capella_query_indexes.new_indexes["idx4"]: Destroying...
couchbase-capella_query_indexes.new_indexes["idx5"]: Destruction complete after 0s
couchbase-capella_query_indexes.new_indexes["idx3"]: Destruction complete after 0s
couchbase-capella_query_indexes.new_indexes["idx2"]: Destruction complete after 0s
couchbase-capella_query_indexes.new_indexes["idx1"]: Destruction complete after 0s
couchbase-capella_query_indexes.new_indexes["idx4"]: Destruction complete after 0s

Apply complete! Resources: 0 added, 0 changed, 5 destroyed.

```

// list indexes

```
terraform plan
╷
│ Warning: Provider development overrides are in effect
│
│ The following provider development overrides are set in the CLI configuration:
│  - couchbasecloud/couchbase-capella in /Users/$USER/GolandProjects/terraform-provider-couchbase-capella/bin
│
│ The behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published releases.
╵
data.couchbase-capella_query_indexes.list: Reading...
couchbase-capella_query_indexes.new_indexes["idx5"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx2"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx3"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx1"]: Refreshing state...
couchbase-capella_query_indexes.new_indexes["idx4"]: Refreshing state...
data.couchbase-capella_query_indexes.list: Read complete after 1s

Changes to Outputs:
  + list_indexes = {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + data            = [
          + {
              + definition = "CREATE INDEX `idx1` ON `test`.`test`.`test`(`field1`)"
              + index_name = "idx1"
            },
          + {
              + definition = "CREATE INDEX `idx2` ON `test`.`test`.`test`(`field2`)"
              + index_name = "idx2"
            },
          + {
              + definition = "CREATE INDEX `idx3` ON `test`.`test`.`test`(`field3`)"
              + index_name = "idx3"
            },
          + {
              + definition = "CREATE INDEX `idx4` ON `test`.`test`.`test`(`field5`)"
              + index_name = "idx4"
            },
          + {
              + definition = "CREATE INDEX `idx5` ON `test`.`test`.`test`(`field5`)"
              + index_name = "idx5"
            },
        ]
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
    }

You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
```

// import index idx1 in test.test.est

```
couchbase-capella_query_indexes.idx1: Importing from ID "index_name=idx1,collection_name=test,scope_name=test,bucket_name=test,cluster_id=ffffffff-aaaa-1414-eeee-000000000000,project_id=ffffffff-aaaa-1414-eeee-000000000000,organization_id=ffffffff-aaaa-1414-eeee-000000000000"...
couchbase-capella_query_indexes.idx1: Import prepared!
  Prepared couchbase-capella_query_indexes for import
couchbase-capella_query_indexes.idx1: Refreshing state...
2024-09-25T12:43:09.513-0700 [WARN]  Provider "registry.terraform.io/couchbasecloud/couchbase-capella" produced an unexpected new value for couchbase-capella_query_indexes.idx1 during refresh.
      - .collection_name: was null, but now cty.StringVal("test")
      - .index_keys: was null, but now cty.ListVal([]cty.Value{cty.StringVal("`c0`")})
      - .index_name: was cty.StringVal("index_name=idx1,collection_name=test,scope_name=test,bucket_name=test,cluster_id=ffffffff-aaaa-1414-eeee-000000000000,project_id=ffffffff-aaaa-1414-eeee-000000000000,organization_id=ffffffff-aaaa-1414-eeee-000000000000"), but now cty.StringVal("idx1")
      - .is_primary: was null, but now cty.False
      - .project_id: was null, but now cty.StringVal("ffffffff-aaaa-1414-eeee-000000000000")
      - .scope_name: was null, but now cty.StringVal("test")
      - .where: was null, but now cty.StringVal("")
      - .bucket_name: was null, but now cty.StringVal("test")
      - .organization_id: was null, but now cty.StringVal("ffffffff-aaaa-1414-eeee-000000000000")
      - .cluster_id: was null, but now cty.StringVal("ffffffff-aaaa-1414-eeee-000000000000")

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
```

// create and build 5 deferred indexes

```
terraform apply
╷
│ Warning: Provider development overrides are in effect
│
│ The following provider development overrides are set in the CLI configuration:
│  - couchbasecloud/couchbase-capella in /Users/hiteshwalia/GolandProjects/terraform-provider-couchbase-capella/bin
│
│ The behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published
│ releases.
╵

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create
 <= read (data resources)

Terraform will perform the following actions:

  # data.couchbase-capella_query_index_monitor.mon_indexes will be read during apply
  # (depends on a resource or a module with changes pending)
 <= data "couchbase-capella_query_index_monitor" "mon_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + indexes         = [
          + "idx1",
          + "idx2",
          + "idx3",
          + "idx4",
          + "idx5",
        ]
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
    }

  # couchbase-capella_query_indexes.build_idx will be created
  + resource "couchbase-capella_query_indexes" "build_idx" {
      + bucket_name     = "test"
      + build_indexes   = [
          + "idx1",
          + "idx2",
          + "idx3",
          + "idx4",
          + "idx5",
        ]
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
    }

  # couchbase-capella_query_indexes.new_indexes["idx1"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field1",
        ]
      + index_name      = "idx1"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build = true
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx2"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field2",
        ]
      + index_name      = "idx2"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build = true
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx3"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field3",
        ]
      + index_name      = "idx3"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build = true
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx4"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field5",
        ]
      + index_name      = "idx4"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build = true
        }
    }

  # couchbase-capella_query_indexes.new_indexes["idx5"] will be created
  + resource "couchbase-capella_query_indexes" "new_indexes" {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + index_keys      = [
          + "field5",
        ]
      + index_name      = "idx5"
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
      + with            = {
          + defer_build = true
        }
    }

Plan: 6 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + watch_indexes = {
      + bucket_name     = "test"
      + cluster_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + collection_name = "test"
      + indexes         = [
          + "idx1",
          + "idx2",
          + "idx3",
          + "idx4",
          + "idx5",
        ]
      + organization_id = "ffffffff-aaaa-1414-eeee-000000000000"
      + project_id      = "ffffffff-aaaa-1414-eeee-000000000000"
      + scope_name      = "test"
    }

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

couchbase-capella_query_indexes.new_indexes["idx1"]: Creating...
couchbase-capella_query_indexes.build_idx: Creating...
couchbase-capella_query_indexes.new_indexes["idx2"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx5"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx3"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx4"]: Creating...
couchbase-capella_query_indexes.new_indexes["idx1"]: Creation complete after 1s
couchbase-capella_query_indexes.new_indexes["idx2"]: Creation complete after 2s
couchbase-capella_query_indexes.new_indexes["idx5"]: Creation complete after 3s
couchbase-capella_query_indexes.new_indexes["idx4"]: Creation complete after 4s
couchbase-capella_query_indexes.new_indexes["idx3"]: Creation complete after 5s
couchbase-capella_query_indexes.build_idx: Still creating... [10s elapsed]
couchbase-capella_query_indexes.build_idx: Still creating... [20s elapsed]
couchbase-capella_query_indexes.build_idx: Still creating... [30s elapsed]
couchbase-capella_query_indexes.build_idx: Still creating... [40s elapsed]
couchbase-capella_query_indexes.build_idx: Still creating... [50s elapsed]
couchbase-capella_query_indexes.build_idx: Still creating... [1m0s elapsed]
couchbase-capella_query_indexes.build_idx: Creation complete after 1m1s
data.couchbase-capella_query_index_monitor.mon_indexes: Reading...
data.couchbase-capella_query_index_monitor.mon_indexes: Still reading... [10s elapsed]
data.couchbase-capella_query_index_monitor.mon_indexes: Still reading... [20s elapsed]
data.couchbase-capella_query_index_monitor.mon_indexes: Still reading... [30s elapsed]
data.couchbase-capella_query_index_monitor.mon_indexes: Still reading... [40s elapsed]
data.couchbase-capella_query_index_monitor.mon_indexes: Still reading... [50s elapsed]
data.couchbase-capella_query_index_monitor.mon_indexes: Still reading... [1m0s elapsed]
data.couchbase-capella_query_index_monitor.mon_indexes: Read complete after 1m1s

Apply complete! Resources: 6 added, 0 changed, 0 destroyed.

Outputs:

watch_indexes = {
  "bucket_name" = "test"
  "cluster_id" = "ffffffff-aaaa-1414-eeee-000000000000"
  "collection_name" = "test"
  "indexes" = toset([
    "idx1",
    "idx2",
    "idx3",
    "idx4",
    "idx5",
  ])
  "organization_id" = "ffffffff-aaaa-1414-eeee-000000000000"
  "project_id" = "ffffffff-aaaa-1414-eeee-000000000000"
  "scope_name" = "test"
}
```

</details>

## Required Checklist:

- [X] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [X] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [X] I have added any necessary documentation (if required)
- [X] I have run make fmt and formatted my code
- [ ] I have made sure that no schema field is marked with both requiresReplace and computed

## Further comments

I left `I have made sure that no schema field is marked with both requiresReplace and computed` unchecked since `scope_name` should be set to `requiresReplace` but since it has default value it must be `computed`

### CODE REVIEW

File: internal/resources/gsi.go [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+					" WITH { \"defer_build\": %t,  \"num_replica\": %d} ",
+					plan.With.DeferBuild.ValueBool(),
+					plan.With.NumReplica.ValueInt64(),
+				)
+
+				ddl += withClause
+			}
+
+		}
+	}
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:127]: In our codebase, we generally do not use many if-else statements; whenever it is required, we use a switch statement instead.

File: internal/resources/gsi.go [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+		//
+		// In this case we can ignore the error.
+		if apiError, ok := err.(*api.Error); ok {
+			if apiError.HttpStatusCode == http.StatusInternalServerError &&
+				strings.Contains(
+					strings.ToLower(apiError.Message), "build already in progress",
+				) {
+				return nil
+			}
+		}
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:432]: Is it possible for this error to occur when other indexes are being built and we are trying to perform another operation? Will it return the same error? If we return nil in this case, does it not give the user the perception that their operation was successful?

File: internal/resources/gsi.go [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+	ddlResponse := api.IndexDDLResponse{}
+	err = json.Unmarshal(response.Body, &ddlResponse)
+	if err != nil {
+		return err
+	}
+
+	//  There are some cases where an operation fails yet query service returns 200 OK.
+	//	For example, when an index is not found or already exists.
+	//  In this case, query service returns errors attribute.
+	//  See MB-62943 for more details.
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:446]: Are these MB publicly accessible?

File: internal/resources/gsi.go [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+func (g *GSI) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !state.BuildIndexes.IsNull() {
+		return
+	}
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:152]: Can you please add a GoDoc comment explaining why we are returning here?

File: internal/api/gsi.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+// IndexDefinitionResponse represents a single index definition.
+type IndexDefinitionResponse struct {
+	Bucket       string   `json:"bucket"`
+	Scope        string   `json:"scope"`
+	Collection   string   `json:"collection"`
+	IsPrimary    bool     `json:"is_primary"`
+	IndexName    string   `json:"indexName"`
+	SecExprs     []string `json:"secExprs"`
```
Comment (aniket-Kumar-c) [inline internal/api/gsi.go:28]: Can you please add a GoDoc comment explaining what secExprs contains?

File: internal/resources/gsi.go [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+				indexName = "#primary"
+			}
+			ddl = fmt.Sprintf(
+				"CREATE PRIMARY INDEX `%s` ON `%s`.`%s`.`%s`  WITH { \"defer_build\": %t,  \"num_replica\": %d }",
+				indexName,
+				plan.BucketName.ValueString(),
+				plan.ScopeName.ValueString(),
+				plan.CollectionName.ValueString(),
+				plan.With.DeferBuild.ValueBool(),
+				plan.With.NumReplica.ValueInt64(),
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:85]: From the schema, it seems that NumReplica and DeferBuild are optional fields, and we haven't populated any default values for them. So, what values are we using if the customer doesn't provide them?

File: internal/resources/gsi.go [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+		return
+	}
+
+	ddl := fmt.Sprintf(
+		"ALTER INDEX `%s` ON `%s`.`%s`.`%s` WITH { \"action\": \"replica_count\", \"num_replica\" : %d }",
+		plan.IndexName.ValueString(),
+		plan.BucketName.ValueString(),
+		plan.ScopeName.ValueString(),
+		plan.CollectionName.ValueString(),
+		plan.With.NumReplica.ValueInt64(),
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:255]: Same as above

From the schema, it seems that NumReplica is optional fields, and we haven't populated any default values for them. So, what values are we using if the customer doesn't provide them?

File: internal/resources/gsi.go [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+func (g *GSI) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !state.BuildIndexes.IsNull() {
+		return
+	}
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:281]: Same as above

Can you please add a GoDoc comment explaining why we are returning here?

File: internal/resources/gsi.go [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+
+	if err := g.executeGsiDdl(ctx, &state, ddl); err != nil {
+		resp.Diagnostics.AddError(
+			"An error occurred while executing index DDL",
+			"Error during index DDL execution: "+err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
```
Comment (aniket-Kumar-c) [inline internal/resources/gsi.go:306]: Is there any specific reason why we are saving back to the state file during the delete operation when no error is encountered?

File: examples/gsi/README.md [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+    {
+      "definition" = "CREATE INDEX `idx5` ON `test`.`test`.`test`(`field5`)"
+      "index_name" = "idx5"
+    },
+  ])
+  "organization_id" = "ffffffff-aaaa-1414-eeee-000000000000"
+  "project_id" = "ffffffff-aaaa-1414-eeee-000000000000"
+  "scope_name" = "test"
+}
+```
```
Comment (aniket-Kumar-c) [inline examples/gsi/README.md:436]: We are still able to see the indexes in output even after deletion. Is this intentional?

File: examples/getting_started/terraform.template.tfvars [hunk 11]
```diff
@@ -192,4 +192,18 @@ aws_config = {
...
   cidr       = "10.0.0.0/16"
+}
+
+bucket_name     = "test"
+scope_name      = "test"
+collection_name = "test"
+
+index_name   = "idx_pe9"
+index_keys   = ["sourceairport", "destinationairport", "stops", "airline", "id", "ARRAY_COUNT(schedule)"]
+partition_by = ["sourceairport", "destinationairport"]
```
Comment (aniket-Kumar-c) [inline examples/getting_started/terraform.template.tfvars:203]: The fields seem to be from the travel-sample dataset, but we have mentioned 'test' in the bucket name. So, we are changing the values before running the getting_started folder

File: internal/resources/gsi.go (hunk 2 above)
Comment (l0n3star) [reply internal/resources/gsi.go:541]: My original intention was to poll the index but this is not a good approach.

Now provider will return an error.

File: internal/resources/gsi.go (hunk 6 above)
Comment (l0n3star) [reply internal/resources/gsi.go:127]: I don't want client to determine the default.  This should be left to the server.

File: internal/resources/gsi.go (hunk 7 above)
Comment (l0n3star) [reply internal/resources/gsi.go:351]: see previous comment

File: internal/resources/gsi.go (hunk 9 above)
Comment (l0n3star) [reply internal/resources/gsi.go]: Good catch.  Should not save to state file.

Addressed in 218633a9d35cfe509d718b3d1994beb6b0162120

File: examples/gsi/README.md (hunk 10 above)
Comment (l0n3star) [reply examples/gsi/README.md]: Good point, that is confusing.  

Removed in 9f656927a9ba035cac03faf46dc1e23b692258f6

File: examples/getting_started/terraform.template.tfvars (hunk 11 above)
Comment (l0n3star) [reply examples/getting_started/terraform.template.tfvars:203]: Bucket name doesn't matter.

File: internal/resources/gsi.go (hunk 3 above)
Comment (l0n3star) [reply internal/resources/gsi.go:551]: Yes

File: internal/resources/gsi.go (hunk 4 above)
Comment (l0n3star) [reply internal/resources/gsi.go:249]: Addressed in 5c6efe534d7761f3957ee6e5d1301148aba9aa2c

File: internal/api/gsi.go (hunk 5 above)
Comment (l0n3star) [reply internal/api/gsi.go]: Addressed in 5c6efe534d7761f3957ee6e5d1301148aba9aa2c

File: internal/resources/gsi.go (hunk 8 above)
Comment (l0n3star) [reply internal/resources/gsi.go:379]: Addressed in 5c6efe534d7761f3957ee6e5d1301148aba9aa2c

**Review by nishant-tripathy — APPROVED** [review]
LGTM

**Pull request merged**